src/Microsoft.AspNetCore.Http.Connections/HttpConnectionContext.cs

```diff
@@ -140,56 +140,86 @@ public void TickHeartbeat()
                     var transportTask = TransportTask ?? Task.CompletedTask;
 
-                    disposeTask = WaitOnTasks(applicationTask, transportTask);
+                    disposeTask = WaitOnTasks(applicationTask, transportTask, closeGracefully);
                 }
             }
             finally
             {
                 Lock.Release();
             }
 
-            try
-            {
-                await disposeTask;
-            }
-            finally
-            {
-                // REVIEW: Should we move this to the read loops?
-
-                // Complete the reading side of the pipes
-                Application?.Input.Complete();
-                Transport?.Input.Complete();
-            }
+            await disposeTask;
         }
 
-        private async Task WaitOnTasks(Task applicationTask, Task transportTask)
+        private async Task WaitOnTasks(Task applicationTask, Task transportTask, bool closeGracefully)
         {
             try
             {
-                await Task.WhenAll(applicationTask, transportTask);
+                // Closing gracefully means we're only going to close the finished sides of the pipe
+                // If the application finishes, that means it's done with the transport pipe
+                // If the transport finishes, that means it's done with the application pipe
+                if (closeGracefully)
+                {
+                    // Wait for either to finish
+                    var result = await Task.WhenAny(applicationTask, transportTask);
+
+                    // If the application is complete, complete the transport pipe (it's the pipe to the transport)
+                    if (result == applicationTask)
+                    {
+                        Transport.Output.Complete(applicationTask.Exception?.InnerException);
+                        Transport.Input.Complete();
+
+                        try
+                        {
+                            // Transports are written by us and are well behaved, wait for them to drain
+                            await transportTask;
+                        }
+                        finally
+                        {
+                            // Now complete the application
+                            Application.Output.Complete();
+                            Application.Input.Complete();
+                        }
+                    }
+                    else
+                    {
+                        // If the transport is complete, complete the application pipes
+                        Application.Output.Complete(transportTask.Exception?.InnerException);
+                        Application.Input.Complete();
+
+                        try
+                        {
+                            // A poorly written application *could* in theory hang forever and it'll show up as a memory leak
+                            await applicationTask;
```

Comment:
Log before and after this. It should be clear to someone looking at the logs when an application is hanging.